### DISCUSSION TITLE
What Determines Which Fonts are Compiled with Sketch?

### DISCUSSION THREAD

**Discussion by Fishbone69** [olikraus/u8g2#2113] [General]
Since I am new to programming microcontrollers, can someone point to a good reference on exactly how the U8g2 fonts are stored in memory when a sketch is compiled?  I know this is trivial to you guys, but I am trying to figure out the following:

1.  When I compile the code is it the u8g2.setFont that determines whether a font is included in the compiled code?
2.  I think I understand the F() macro function, but essentially it offloads RAM by having the code access it from PROGMEM?
3. If I am going to use a combinations of text and "icons" would it be better to create a new font set and include bitmaps of my icons or should I define only the bitmaps I intend to use and use the drawXMB command?

Thank you, smart people!

Fish

**Comment by olikraus** [comment]
1. Yes, together with a feature called link time optimization (LTO)
2. Yes, but the F macro is only the Arduino implementation on top of the GCC AVR progmem Features. U8g2 Fonts are also stored in progmem, but use a more native approach.
3. Any approach is fine, use the description for drawXBMP to see how to put a bitmap into progmem. Also note that there are many icon fonts included with u8g2. Maybe you can use some of them.

**Comment by Fishbone69** [comment]
I know this is a bit stale, but I am just getting to loading some bitmaps into my code and am still wondering if I should use drawXBM or drawXBMP.  I think I understand the difference...so is is the tradeoff drawXBMP uses program memory so it is slower than drawXBM?

**Comment by olikraus** [comment]
The difference only applies to AVR uC. Most notable difference is RAM consumption. And yes, it might be slower.
